src/evo/deterministicmns.cpp

```diff
@@ -1422,13 +1426,25 @@ bool CheckProUpServTx(CDeterministicMNManager& dmnman, const CTransaction& tx, g
     }
 
     auto mnList = dmnman.GetListForBlock(pindexPrev);
-    auto mn = mnList.GetMN(opt_ptx->proTxHash);
-    if (!mn) {
+    auto dmn = mnList.GetMN(opt_ptx->proTxHash);
+    if (!dmn) {
         return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-protx-hash");
     }
 
+    const bool is_v23_active{DeploymentActiveAfter(pindexPrev, Params().GetConsensus(), Consensus::DEPLOYMENT_V23)};
+
+    // Don't allow legacy scheme versioned transactions after upgrading to basic scheme
+    if (is_v23_active && dmn->pdmnState->nVersion >= ProTxVersion::BasicBLS && opt_ptx->nVersion == ProTxVersion::LegacyBLS) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-protx-version-downgrade");
+    }
+
+    // Nodes using the legacy scheme must first upgrade to the basic scheme before upgrading further
+    if (is_v23_active && dmn->pdmnState->nVersion == ProTxVersion::LegacyBLS && opt_ptx->nVersion > ProTxVersion::BasicBLS) {
+        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-protx-version-upgrade");
+    }
```

Comment:
nit: I'd prefer a larger if block for `is_v23_active` to avoid duplicate conditions

Reply:
See [comment](https://github.com/dashpay/dash/pull/6665#discussion_r2159738632)